submissions/misshatsker/object-oriented-js/app.js

```diff
@@ -0,0 +1,171 @@
+var leftSide = 0;
+var rightSide = 500;
+var topSide = 0;
+var bottomSide = 500;
+var initialGroundY = 60;
+
+var Enemy = function(x, y, speed) {
+    this.sprite = 'images/enemy-bug.png';
+    this.speed = speed;
+    this.x = x;
+    this.y = y;
+};
+
+Enemy.prototype.update = function(dt) {
+    var shift = this.speed * dt;
+    this.x = this.x + shift;
+    if (this.x >= rightSide) {
+        this.x = -this.width;
+    }
+};
+
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.width = 80;
+Enemy.prototype.height = 70;
+
+var Player = function(x, y, speed) {
+    this.sprite = 'images/char-princess-girl.png';
+    this.speed = speed;
+    this.x = x;
+    this.y = y;
+};
+
+Player.prototype.width = 70;
+Player.prototype.height = 70;
+
+Player.prototype.update = function() {
+    const isOverlapDetected = allEnemies.some((enemy) => {
+        return checkIsOverlap(
+            {
+                x: player.x,
+                y: player.y
+            },
+            {
+                x: player.x + player.width,
+                y: player.y + player.height
+            },
+            {
+                x: enemy.x,
+                y: enemy.y
+            },
+            {
+                x: enemy.x + enemy.width,
+                y: enemy.y + enemy.height
+            }
+        );
+    });
+
+    if (isOverlapDetected) {
+        gameOver(false);
+    }
+}
+
+Player.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+}
+
+Player.prototype.handleInput = function(action) {
+    switch (action) {
+        case 'left': {
+            if (this.x > leftSide) {
+                this.x -= this.speed;
+            }
+            break;
+        }
+        case 'right': {
+            if (this.x + this.width + this.speed < rightSide) {
+                this.x += this.speed;
+            }
+            break;
+        }
+        case 'down': {
+            if (this.y + this.height + this.speed < bottomSide) {
+                this.y += this.speed;
+            }
+            break;
+        }
+        case 'up': {
+            this.y -= this.speed;
+            if (this.y <= 0) {
+                gameOver(true);
+            }
+
+            break;
+        }
+    }
+}
+
+Player.prototype.setPosition = function(x, y) {
+    this.x = x;
+    this.y = y;
+}
+
+function getRandomNumberFromARange(min, max) {
+    return (Math.random() * (max - min)) + min;
+}
+
+function createRandomEnemy(row) {
+    var x = getRandomNumberFromARange(-400, -100);
+    var y = initialGroundY + row * 80;
+    var speed = getRandomNumberFromARange(100, 300);
+
+    return new Enemy(x, y, speed);
+}
+
+var allEnemies = [
+    createRandomEnemy(0),
+    createRandomEnemy(1), 
+    createRandomEnemy(2)
+];
+
+var initialPlayerX = 200;
+var initialPlayerY = 400;
+var playerSpeed = 75;
+var player = new Player(initialPlayerX, initialPlayerY, playerSpeed);
```

Comment:
Requirements address issue we have in this fragment and in many places across the code.
Please find it, post here as a comment and fix code accordingly.

Reply:
My fault. I should have used `const`s/`let`s instead of `var`s. Not for an excuse but I've been misled by using var in the original repo.